status/condition_set.go

```diff
@@ -104,18 +104,27 @@ func (c ConditionSet) IsTrue(conditionTypes ...string) bool {
 func (c ConditionSet) Set(condition Condition) (modified bool) {
 	conditionType := condition.Type
 	var conditions []Condition
-	for _, c := range c.object.GetConditions() {
-		if c.Type != conditionType {
-			conditions = append(conditions, c)
+	for _, cond := range c.object.GetConditions() {
+		if cond.Type != conditionType {
+			conditions = append(conditions, cond)
 		} else {
-			// If we'd only update the LastTransitionTime, then return.
-			condition.LastTransitionTime = c.LastTransitionTime
-			if reflect.DeepEqual(condition, c) {
-				return false
+			condition.LastTransitionTime = cond.LastTransitionTime
+			condition.ObservedGeneration = cond.ObservedGeneration
```

Comment:
This won't update the observed generation when the condition is set for the first time. We should ideally update the observed generation always in the same spot that we were updating the time below so that we know that we always set it. There's probably some handling that we'll have to have so that we are still updating the ObservedGeneration but only update TransitionTime when there is a status update

Reply:
I also have this feeling that there is a better way to factor this so that we don't have so many conditionals. Generally, prefer short-circuiting when you can if it's possible based on the behavior that we want here.

Reply:
So, a couple things:
1. I still felt like we could factor this better so I tried my hand at it for a second. Does this work?

```go
// Set sets or updates the Condition on Conditions for Condition.Type.
// If there is an update, Conditions are stored back sorted.
func (c ConditionSet) Set(condition Condition) (modified bool) {
	conditionType := condition.Type
	var conditions []Condition
	for _, cond := range c.object.GetConditions() {
		if cond.Type != conditionType {
			conditions = append(conditions, cond)
		} else {
			condition.LastTransitionTime = metav1.Now()
			condition.ObservedGeneration = c.object.GetGeneration()
			if condition.Status == cond.Status {
				condition.LastTransitionTime = cond.LastTransitionTime
			}
			if reflect.DeepEqual(condition, cond) {
				return false
			}
		}
	}
	conditions = append(conditions, condition)
	// Sorted for convenience of the consumer, i.e. kubectl.
	sort.Slice(conditions, func(i, j int) bool { return conditions[i].Type < conditions[j].Type })
	c.object.SetConditions(conditions)

	// Recompute the root condition after setting any other condition
	c.recomputeRootCondition(conditionType)
	return true
}
```

2. I wanted to validate when the observedGeneration of the root condition should be updated. Is it set to the minimum of all dependent status conditions or is it set the whatever the current generation is? My gut would be that this would only be meaningful when it is set to the minimum -- that way you know that the current state of Readiness reflects an old view of state, but I need to check this in upstream.

Reply:
```
if condition.Status == cond.Status {
				condition.LastTransitionTime = cond.LastTransitionTime
			}
```
It may be true/enough for our case, but the original comment, actually update the LastTransitionTime when any of Status/Reason/Message, so just checking Status changes the original logic how LastTransitionTime gets updated

Reply:
This does the same, right? Basically, the check is: Did the status change? If it did, then the transition time should update. We shouldn't update the transition time if the reason updated, but we should still update the reason on the condition if there was a request to do so.

Reply:
> I wanted to validate when the observedGeneration of the root condition should be updated

On this point: Thinking through it -- I think the correct behavior here is to update the root condition to Unknown if the observedGeneration doesn't match across. This is the known state of the root condition: We don't know whether it will be true or false since the dependent conditions haven't had the chance to catch-up yet. Once all of the status conditions match their observed generation to the latest, the root condition can update to a known state (true/false)

Reply:
This also means that a constantly updating spec will never resolve into a "known" state for the Ready condition but I see this as expected since adding the observedGeneration to ready means that the Ready condition that we have should reflect the actual state at that generation. If we don't know it because we were in a state of transitioning, then we have to report that.